compiler/src/typed/parmatch.re

```diff
@@ -699,6 +699,8 @@ let full_match = (closing, env) =>
       /* extensions */
       List.length(env) == c.cstr_consts + c.cstr_nonconsts;
     }
+  | [({pat_desc: TPatConstant(Const_bool(_))}, _), ..._] =>
+    List.length(env) == 2
```

Comment:
Why is it okay that we just check the env has a length of 2?

Reply:
The description of this function explains the first part better than I can, so let's start there:
> Check whether the first column of env makes up a complete signature or not. We work on the discriminating patterns of each sub-matrix: they are simplified, and are not omega/Tpat_any. 

Essentially `env` contains all of the unique non-"any " patterns for a section of the total match. Since we know that the type is boolean, there aren't any "any" patterns, and they're all unique, if the env contains two patterns, they must be the patterns `true` and `false`.

A similar technique is used above for variants—as long as the number of patterns in the env is equal to the total number of constructors, since they're simplified and unique, that means each constructor appears exactly once.

Hope that makes sense!

Reply:
💯